crates/astria-sequencer/src/service/mempool/mod.rs

```diff
@@ -173,133 +291,91 @@ async fn handle_check_tx<S: accounts::StateReadExt + address::StateReadExt + 'st
-        match removal_reason {
-            RemovalReason::Expired => {
-                metrics.increment_check_tx_removed_expired();
-                return response::CheckTx {
-                    code: Code::Err(AbciErrorCode::TRANSACTION_EXPIRED.value()),
-                    info: "transaction expired in app's mempool".into(),
-                    log: "Transaction expired in the app's mempool".into(),
-                    ..response::CheckTx::default()
-                };
-            }
-            RemovalReason::FailedPrepareProposal(err) => {
-                metrics.increment_check_tx_removed_failed_execution();
-                return response::CheckTx {
-                    code: Code::Err(AbciErrorCode::TRANSACTION_FAILED.value()),
-                    info: "transaction failed execution in prepare_proposal()".into(),
-                    log: format!("transaction failed execution because: {err}"),
-                    ..response::CheckTx::default()
-                };
-            }
-            RemovalReason::NonceStale => {
-                return response::CheckTx {
-                    code: Code::Err(AbciErrorCode::INVALID_NONCE.value()),
-                    info: "transaction removed from app mempool due to stale nonce".into(),
-                    log: "Transaction from app mempool due to stale nonce".into(),
-                    ..response::CheckTx::default()
-                };
-            }
-            RemovalReason::LowerNonceInvalidated => {
-                return response::CheckTx {
-                    code: Code::Err(AbciErrorCode::LOWER_NONCE_INVALIDATED.value()),
-                    info: "transaction removed from app mempool due to lower nonce being \
-                           invalidated"
-                        .into(),
-                    log: "Transaction removed from app mempool due to lower nonce being \
-                          invalidated"
-                        .into(),
-                    ..response::CheckTx::default()
-                };
-            }
-        }
-    };
+    // note: decide if worth moving to post-insertion, would have to recalculate cost
+    metrics.record_transaction_in_mempool_size_bytes(tx_len);
 
-    let finished_check_removed = Instant::now();
-    metrics.record_check_tx_duration_seconds_check_removed(
-        finished_check_removed.saturating_duration_since(finished_check_chain_id),
-    );
+    Ok(signed_tx)
+}
+
+/// Attempts to insert the transaction into the mempool.
+///
+/// Returns a `Err(response::CheckTx)` with an error code and message if the transaction fails
+/// insertion into the mempool.
+#[expect(clippy::too_many_lines, reason = "error handling is long")]
+async fn insert_into_mempool<S: accounts::StateReadExt + address::StateReadExt + 'static>(
+    mempool: &AppMempool,
+    state: &S,
+    signed_tx: SignedTransaction,
+    metrics: &'static Metrics,
+) -> Result<(), response::CheckTx> {
```

Comment:
I have noted it for `RemovalReason`, but it also applies here: all the error-to-response conversions could be made a lot neater by adding a trait `IntoCheckTxResponse` and moving the conversions there instead of having them inlined in the function body.

Reply:
Nice. Wrote the trait and defined if for `RemovalReason` and `InsertionError`

Reply:
Yeah, neat.